service/storage_service.cc

```diff
@@ -858,6 +858,7 @@ future<> storage_service::topology_transition(state_change_hint hint) {
     SCYLLA_ASSERT(this_shard_id() == 0);
     co_await topology_state_load(std::move(hint)); // reload new state
 
+    _topology_state_machine.reload_count++;
```

Comment:
Is this the only place we broadcast? Can we guarantee that it will always be the only place? Looks like a footgun, someone might call `broadcast` but forget about `reload_count`. It looks like this should be part of a larger abstraction behind interface that calls `broadcast` and increments counter together,  perhaps `condition_variable` itself

Reply:
It has nothing to do with broadcast but state reload. It is an optimistic locking pattern where you access the state without locking and discard the result if you detected that the state changes.

Reply:
The problem: someone might easily repeat the exact same mistake in the future. My proposal: make it future-proof somehow. Perhaps by wrapping `condition_variable` in a higher-level abstraction and using it instead. Or extending `condition_variable` API with `wait(predicate)` where `predicate` returns a `future<>`, the API would provide atomicity guarantee, it seems basically what you're doing here but "manually"

Reply:
> The problem: someone might easily repeat the exact same mistake in the future. My proposal: make it future-proof somehow. Perhaps by wrapping `condition_variable` in a higher-level abstraction and using it instead. Or extending `condition_variable` API with `wait(predicate)` where `predicate` returns a `future<>`, the API would provide atomicity guarantee, it seems basically what you're doing here but "manually"

You are too hang up on "future-proof " which somehow magically will make people to not make bugs. Even if you had an API you described nobody guaranties it will be used when needed.

Anyway I do not see how API you describe will work and how it will solve the problem. You are welcome to add the API and we will discuss it during review.

Meanwhile, if you want easily backportable fix, this is the fix.

Reply:
There seems to be the exact same order of calls in here:
https://github.com/scylladb/scylladb/blob/a49ed7074d94fbe74213d71099897edf6d3f72ac/service/raft/group0_state_machine.cc#L315-L321
so maybe we'd want to put the `reload_count` increment there as well?

Plus there is another call of `_topology_state_machine.event.broadcast()` in here:
https://github.com/scylladb/scylladb/blob/a49ed7074d94fbe74213d71099897edf6d3f72ac/service/storage_service.cc#L857-L862
so maybe also there - unless the `reload_count` applies just to the `topology_state_load()`, in which case it would make sense to at least put it inside of the `topology_state_load()` to handle the 2 places 🤔